### PR TITLE
feat(rules): add TypeScript MCP command injection and SSRF detection

### PR DESCRIPTION
**Pull request by Tito0015**
## Summary
Adds TypeScript MCP security rules for command injection (CWE-78) and SSRF (CWE-918), aligned with the hardened patterns from [semgrep/semgrep-rules#4052](https://github.com/semgrep/semgrep-rules/pull/4052).

- `rules/ts-mcp-command-injection/` — taint from MCP tool/request handlers to shell execution sinks (including `execFile` with `shell: true`)
- `rules/ts-mcp-ssrf/` — taint to `fetch` and `axios` HTTP methods

Note: These rules were originally authored by me for the Semgrep community registry (PR #4052). This change ports the same detection logic into the Aikido Opengrep rules layout.

## Validation
- `opengrep validate` on both rule YAML files
- `opengrep --test rules/ts-mcp-command-injection/` — 1 rule, 0 failures
- `opengrep --test rules/ts-mcp-ssrf/` — 1 rule, 0 failures

### CODE REVIEW

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import * as cp from "child_process";
+import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import { z } from "zod";
+
+const server = new McpServer({ name: "test-server", version: "1.0.0" });
+
+server.setRequestHandler("tools/call", async (REQ) => {
+  const cmd = REQ.params.arguments.command;
+  // ruleid: ts-mcp-command-injection
+  exec(cmd);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts:11]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  return {};
+});
+
+server.tool("run-command", { command: z.string() }, async (args) => {
+  // ruleid: ts-mcp-command-injection
+  exec(args.command, () => {});
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts:17]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  return { content: [{ type: "text" as const, text: "done" }] };
+});
+
+server.tool("run-sync", { command: z.string() }, async ({ command }) => {
+  // ruleid: ts-mcp-command-injection
+  execSync(command);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts:23]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  return { content: [{ type: "text" as const, text: "ok" }] };
+});
+
+server.registerTool(
+  "run-v2",
+  { description: "run", inputSchema: z.object({ command: z.string() }) },
+  async ({ command }) => {
+    // ruleid: ts-mcp-command-injection
+    cp.execSync(command);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts:32]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    return { content: [{ type: "text" as const, text: "done" }] };
+  }
+);
+
+server.tool("safe-zod", { command: z.string() }, async ({ command }) => {
+  const safe = z.string().parse(command);
+  // ok: ts-mcp-command-injection
+  exec(safe);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts:40]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

**Comment by Tito0015** [comment]
Thanks to the Aikido Deep Review bot for the sharp catches! The rule has been updated to cover all callback overloads and strict sanitization edge cases.

I originally authored this pattern for the Semgrep registry (PR #4052), but wanted to port it natively to Opengrep to support the Aikido ecosystem with high-precision threat detections.

Through my project, Veraptos, I’ve built a deterministic CPG meta-compiler that synthesizes engine-native YAML rules directly from data-flow traces—focusing on low-noise detection with bounded structural candidates (< 10 findings per scan) rather than relying on probabilistic LLM generation.

I'm currently compiling a suite of rules targeting emerging attack surfaces (MCP server handlers, supply-chain vectors, and Node/TS framework patterns). I'd be glad to submit a batch of 5–10 of these directly to opengrep-rules if the team finds this format valuable.

Let me know if the maintainers are open to a quick technical chat on the pipeline or feedback on incoming rule sets!

**Comment by Tito0015** [comment]
Note for reviewers: The `Aikido Security: check code` CI check failed because it successfully detected the intentional vulnerabilities (`eval`, `axios`, `fetch`) inside the `.ts` test fixtures. All local Opengrep validation and `--test` suites are passing completely.

Also, just a quick heads-up that these exact hardened rules and test fixtures were just approved for merge by the core maintainers upstream in the official `semgrep/semgrep-rules` repository (PR #4052). I proactively update it. I remain on standby if you need anything.

File: rules/ts-mcp-command-injection/ts-mcp-command-injection.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  async ({ command }) => {
+    // ruleid: ts-mcp-command-injection
+    cp.execSync(command);
+    return { content: [{ type: "text" as const, text: "done" }] };
+  }
+);
+
+server.tool("eval-concat", { command: z.string() }, async ({ command }) => {
+  // ruleid: ts-mcp-command-injection
+  eval("prefix" + command + "suffix");
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-command-injection/ts-mcp-command-injection.ts]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-ssrf/ts-mcp-ssrf.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+server.setRequestHandler("resources/read", async (REQ) => {
+  const url = REQ.params.arguments.url;
+  // ruleid: ts-mcp-ssrf
+  await fetch(url);
+  return {};
+});
+
+server.tool("fetch-url", { url: z.string() }, async ({ url }) => {
+  // ruleid: ts-mcp-ssrf
+  await axios.get(url);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-ssrf/ts-mcp-ssrf.ts]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-ssrf/ts-mcp-ssrf.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  async ({ url }) => {
+    // ruleid: ts-mcp-ssrf
+    await fetch(url);
+    return { content: [{ type: "text" as const, text: "done" }] };
+  }
+);
+
+server.tool("patch-url", { url: z.string() }, async ({ url }) => {
+  // ruleid: ts-mcp-ssrf
+  await axios.patch(url);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-ssrf/ts-mcp-ssrf.ts]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.

File: rules/ts-mcp-ssrf/ts-mcp-ssrf.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import { z } from "zod";
+import axios from "axios";
+
+const server = new McpServer({ name: "test-server", version: "1.0.0" });
+
+server.setRequestHandler("resources/read", async (REQ) => {
+  const url = REQ.params.arguments.url;
+  // ruleid: ts-mcp-ssrf
+  await fetch(url);
```
[the remark this answers is not in the document]
Comment (Tito0015) [reply rules/ts-mcp-ssrf/ts-mcp-ssrf.ts]: @AikidoSec ignore: Intentional vulnerable code for rule test fixture.
